04-onion-routing.md

```diff
@@ -367,6 +379,152 @@ otherwise meets the amount criterion (eg. some other failure, or
+- If `encrypted_data` contains a `next_blinding_override`:
+  - MUST use it as the next blinding point
+- Otherise:
+  - MUST use `E(1)` as the next blinding point
+- MUST forward the onion and communicate the next blinding point to the next node
+
+An intermediate node in the blinded route:
+
+- MUST compute:
+  - `ss(i) = SHA256(k(i) * E(i))` (standard ECDH)
+  - `b(i) = HMAC256("blinded_node_id", ss(i)) * k(i)`
+  - `rho(i) = HMAC256("rho", ss(i))`
+  - `E(i+1) = SHA256(E(i) || ss(i)) * E(i)`
+- MUST use `b(i)` instead of its private key `k(i)` to decrypt the onion. Note
+  that the node may instead tweak the onion ephemeral key with
+  `HMAC256("blinded_node_id", ss(i))` which achieves the same result.
+- If an `encrypted_data` field is provided:
+  - MUST decrypt it using `rho(i)` and `E(i)`
+  - MUST use the decrypted fields to locate the next node
+- If `encrypted_data` contains a `next_blinding_override`:
+  - MUST use it as the next blinding point
+- Otherise:
+  - MUST use `E(i+1)` as the next blinding point
+- MUST forward the onion and communicate the next blinding point to the next node
+
+The final recipient:
+
+- MUST compute:
+  - `ss(r) = SHA256(k(r) * E(r))` (standard ECDH)
+  - `b(r) = HMAC256("blinded_node_id", ss(r)) * k(r)`
+  - `rho(r) = HMAC256("rho", ss(r))`
+- If an `encrypted_data` field is provided:
+  - MUST decrypt it using `rho(r)` and `E(r)`
+- MUST ignore the message if `E(r)` or `path_id` does not match the
+  blinded route it created
+
+#### Rationale
+
+Route blinding is a lightweight technique to provide recipient anonymity.
+It's more flexible than rendezvous because it lets senders choose what data
+they include in each node's onion payload and retry with updated values if
+the route fails.
+
+Use-cases where route blinding is useful include:
+
+- Recipient anonymity when receiving payments
+- Using unannounced channels in invoices without revealing them
+- Forcing a message to go through a specific set of intermediaries that can witness it
+- Providing anonymous reply paths for onion messages (offers, stuckless payments, etc)
+
+Each node in the blinded route needs to receive `E(i)` to be able to decrypt
+the onion and the `encrypted_data`. Protocols that use route blinding must
+specify how this value is propagated between nodes.
+
+When concatenating two blinded routes generated by different nodes, the
+last node of the first route needs to know the first `blinding` of the second
+route: the `next_blinding_override` field must be used to transmit this data.
+
+The final recipient must verify that the blinded route is used in the right
+context (e.g. for a specific payment). It can do so by storing `E(r)` and the
+context (e.g. a `payment_hash`), and verifying that they match when receiving
+the onion. Otherwise, to avoid additional storage cost, it can put some context
+information in the `path_id` field (e.g. a `payment_hash`) and verify that when
+receiving the onion.
```

Comment:
No, this is wrong.  `path_id` *MUST* be a secret unique to the invoice (or whatever she presented the blinded route inside).

Consider: Alice anonymously presents invoice containing a blinded route, presents it to mallory.

If it's just the payment_hash, Mallory can create his own encrypted_data_tlvencrypted_data_tlv with that `path_id` to Alice, and pay the invoice.  He then knows that Alice is actually the anon recipient!

Alice knows by the path_id that the payment used her path.  She should treat all unknown path_ids as public, and she can generate path_id from H(secret || payment_hash) ofc.

Reply:
You're right, that comment is a left-over from a while ago, and it's incorrect!
I'll fix that.

Reply:
Fixed in https://github.com/lightning/bolts/pull/765/commits/a9e73be6b8bd86aa938318e3a7fc3370c5be675b